### DISCUSSION TITLE
Hardware

### DISCUSSION THREAD

**Discussion by 14yannick** [Gifford47/HCPBridgeMqtt#53] [Show and tell]
Lets start a thread to talk about Hardware.

**Comment by hardysim** [comment]
I want to add some more sensors to the [Tynet RS485 Bridge](https://shop.tynet.eu/rs485-bridge-diy-hoermann-mqtt-adapter-esp32-s3-dev-board) but I'm a total newbie and [struggle with the correct pin connection](https://community.home-assistant.io/t/hormann-garage-door-via-mqtt/279938/431?u=hardysim) as well on how to bake everything into a new firmware.

Following @14yannick [post at the HA forum](https://community.home-assistant.io/t/hormann-garage-door-via-mqtt/279938/432?u=hardysim), I may use the temp and tumidity sensor BMP280 and for Proximity a hc-sr04.

Can someone post a tutorial  or some pictures to get started?

**Reply by 14yannick** [reply]
Hi Hardysim, 
for connection of the sensor your board has

- exp1 with gnd 3.3v, GPI05 and GPI06. This arrangement make them perfect to connect the BME280. Connect 3.3 to vcc, gnd to gnd, GPIO5 to SCL and GPIO6 to SDA.
- exp2 has GPIO7 and GPIO8 next to power and ground. connect GPIO7 to Trigg and GPIO8 to echo. Please note that you need and and hc-sr04**P** as it has to operate in 3.3v

Once everything is connected enter the ip Address of your HCPBridge in your browser (if you don't know the ip of it look on your router for a client with the name "HCPBRIDGE"). 
then go to the press on show expert configuration (if you don't see it you need to update your firmware first) and adjust the the pin definition of  I2C SDA 6, I2C SCL 5, SR04 trigger 7, SR04 echo 8 according the connection made. Press save and restart.

Then you need a new Firmware add /update to your HCPBridge ip in your browser. OTA user is admin and Password is also admin.
Take the binary from [here]([HCPBridge_0.0.6.1.zip](https://github.com/Gifford47/HCPBridgeMqtt/releases/download/0.0.6.1/HCPBridge_0.0.6.1.zip)). as you have an tynet board use the one in the folder esp32s3_tynet_BME280_HCSR04 or esp32s3_tynet_BME280.

Then you just need to restart your Hörmann garage door and your are done.

**Reply by hardysim** [reply]
Hey @14yannick that's already a great tutorial, thank you very much. I understand that I need to connect the two sensors to the specified pins on my tynet board and update the mapping in the expert > pin definition section on the web ui (and to use the variants of the sensors which use 3.3V instead of 5 - good hint).

Also I need to know a simple way of connecting the sensors - do I need to solder or is there a different way?

I also found the firmware from the releases page and in the zip file folder `esp32s3_tynet_BME280_HCSR04`. I'll upload this as the "firmware" (not the "filesystem") via the OTA page, right?

Will all configuration survive the update? Or do I need to re-set the pin mapping again? Or is the configuration stored separately (along with the wifi, mqtt settings etc.)?

**Reply by 14yannick** [reply]
I think you will not get arround some easy soldering. I would suggest you to put some male Pinheader then you use a dupont female female cable to attache your sensor.
![image](https://github.com/Gifford47/HCPBridgeMqtt/assets/6718914/2550a8a5-0c9d-4aa7-9ff5-4ca323965f60)

Yes upload it as firmware. The configuration should survice the update. Once you have update the firmware go basic settings and adjust the Device name to something like "Garage Door" and save your settings. Due to a bug this value will probably be shown as undefined.

**Reply by jumahrauf** [reply]
I followed all of your instructions to connect my BME280 to my TynetBoard 1.3
I connected the sensor and flashed the firmware. After that I adjusted the pin definition.

Now the values which are shown in HomeAssistant are for all 3 Sensors „-100“. Did I set something wrong?

**Comment by jarpatus** [comment]
Is possible to use DHT22 with Tynet board? There were no section for such configuration in platformio.ini and when I tried to add it myself and compile, resulting fimrware did not boot.

Askong because I just happen to have DHT22 lying around and why not to use it...

**Reply by 14yannick** [reply]
Did you define the env like this.

[env:esp32s3_tynet_dht22]
extends = env:esp32s3_tynet
lib_deps = ${env:esp32.lib_deps} ${common.dht22libs}
build_flags = ${env:esp32.build_flags} -D SENSORS -D USE_DHT22

I have no tynet or s3 board but I was able to use a dht22 with an older esp32E board. In the config file the pin DHT_VCC_PIN is set to 27 so you also have to adjust this in the expert config with the pin you attached your sensor. Don't forget to put a 10k resistor between the dht vcc pin and the 3.3v then it should works.

Please give feedback if it worked so I will upload the build to the release.

**Reply by jarpatus** [reply]
Yes this is what I tried and now re-tried but no luck, after I upload esp32s3_tynet_dht22 firmware then board won't boot (or at least when I run pio .. --target monitor I won't get anything unlike with esp32s3_tynet).

However, I did not connect DHT22 sensor yet, is hardware connection required for board to boot and connect to wifi etc.?

**Reply by 14yannick** [reply]
I tried it as I received my M5Stack Atom S3 lite. I had initially also some issue uploading the firmware. The esp didn't boot and trowing massif error in the Serial. I have reset it with the command below
`python3 -m esptool --chip esp32-s3 erase_flash`
Upload again without any change and it worked also if the the DHT22 is not connected. I just show invalid temp and humidity in Home assistant.

**Reply by jarpatus** [reply]
Ok so I was able to make it work. Erasing flash did not help as I had tried it before but I proceeded then to add lot's of serial debugging to the code and found out that board will reboot when DHT22 sensors are being initialized. So I realized that default value of DHT_VCC_PIN could overlap with something else and I fixed that, erased flash etc. and wham, it works (I thought I could fix value from UI later but no).

While at it, I think DHT22 initialization code is wrong. Whole DHT_VCC_PIN term is incorrect as it should rather be DHT_DATA_PIN and also subsequently pin should not be set as output pin nor set to high as dht->begin() should do initialization (and perhaps should not be called on every sensor read). I fixed this, actually not sure if this was also needed to make DHT22 work, did not try without.

While at it I realized Tynet board has 4 controllable leds so added code to lit leds when board has booted, network has connected and mqtt has connected. Helps debugging a lot as sometimes board cannot connect to wifi or so.

Also I modified MQTT config messages to include device class, unit of measurement etc. so that home assistant will show sensor values correctly without custom YAML.

Should I create pull request(s) for these changes, does somebody want to take a look (disclaimer: I am high level programming language developer and do not know anything about embedded boards so YMMV)?

Edit: For the record, if DHT22 is connected to EXP2 header then correct value for DHT_VCC_PIN would be 7. Values will match with IO# values from Tynet webshop.

**Reply by 14yannick** [reply]
Yes pull request are we
Welcome please make it against the dev Branch. We working on a big redesign of the code to avoid having separate build by sensor.

**Comment by Saicopate** [comment]
Hello, do you happen to know what should be the configuration entries for AHT20 + BMP280 sensor connected over I2C? I have a Tynet board v1.3

**Reply by 14yannick** [reply]
You can connect it to the exp1 and flash the tynet BME Build. But you will only have the readings of the BME as only this sensor is implemented see [here](https://github.com/Gifford47/HCPBridgeMqtt/blob/edda64c082be56d0efae877fe3333b2a830f681d/HCPBridgeESP32/src/main.cpp#L736). As I found [here](https://community.home-assistant.io/t/aht20-bmp280-combined-sensor/366411/7) the AHT20 is on another address. It's a bit the problem of this code base each sensor has to be implemented separately and if add to the mqtt message,... and so one. I started to make a rework of the code base but I have the feeling to reinvent the wheel as so much Boilerplate code is needed just for the mqtt setup, userPreference and sensorReading. So will get a try to the esphome port of this [project](https://github.com/mapero/esphome-hcpbridge).

**Reply by Saicopate** [reply]
Ok, I had a look into main.cpp and I get what you mean. AHT20 would need to be implemented in several places. I am not into C++ so I will have to let it go and use BME sensor only. Thanks for looking into this Yannick!
